buffer/src/test/java/io/netty/buffer/PooledByteBufAllocatorTest.java

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright 2015 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.buffer;
+
+import io.netty.util.internal.SystemPropertyUtil;
+import org.junit.Test;
+
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.CountDownLatch;
+
+public class PooledByteBufAllocatorTest {
+
+    private static final int[] ALLOCATION_SIZES = new int[16 * 1024];
+    static {
+        for (int i = 0; i < ALLOCATION_SIZES.length; i++) {
+            ALLOCATION_SIZES[i] = i;
+        }
+    }
+
+    @Test
+    public void testConcurrentUsage() throws Throwable {
+        long runningTime = SystemPropertyUtil.getLong(
+                "io.netty.buffer.PooledByteBufAllocatorTest.testConcurrentUsageTime", 30000);
+
+        // We use no caches and only one arena to maximize the chance of hitting the race-condition we
+        // had before.
+        ByteBufAllocator allocator = new PooledByteBufAllocator(true, 1, 1, 8192, 11, 0, 0, 0);
+        List<AllocationThread> threads = new ArrayList<AllocationThread>();
+        try {
+            for (int i = 0; i < 512; i++) {
+                AllocationThread thread = new AllocationThread(allocator);
+                thread.start();
+                threads.add(thread);
+            }
+
+            long sleepTime = 500;
+            long startTime = System.currentTimeMillis();
+            long endTime = startTime + runningTime;
+            for (; startTime < endTime; startTime += sleepTime) {
```

Comment:
Why are we vulnerable to overflow ? If we not have enough left after adding stuff to to System.currentTimeMillis() I think we have bigger problems.